Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/internal/GarlandHeckbert_policy_base.h

```diff
@@ -0,0 +1,250 @@
+//
+// $URL$
+// $Id$
+// SPDX-License-Identifier: GPL-3.0-or-later OR LicenseRef-Commercial
+//
+// Author(s)     : Baskin Burak Senbaslar,
+//                 Mael Rouxel-Labbé,
+//                 Julian Komaromy
+
+#ifndef CGAL_SURFACE_MESH_SIMPLIFICATION_POLICIES_INTERNAL_GARLANDHECKBERT_POLICIES_BASE_H
+#define CGAL_SURFACE_MESH_SIMPLIFICATION_POLICIES_INTERNAL_GARLANDHECKBERT_POLICIES_BASE_H
+
+#include <CGAL/license/Surface_mesh_simplification.h>
+
+#include <CGAL/Surface_mesh_simplification/Policies/Edge_collapse/Edge_profile.h>
+#include <CGAL/Surface_mesh_simplification/internal/Common.h>
+#include <CGAL/Surface_mesh_simplification/Policies/Edge_collapse/internal/GarlandHeckbert_core.h>
+
+#include <iostream>
+#include <CGAL/tags.h>
+
+#include <Eigen/Dense>
+
+#include <boost/optional/optional.hpp>
+
+namespace CGAL {
+namespace Surface_mesh_simplification {
+namespace internal {
+  
+template<typename Mat_4>
+Mat_4 combine_matrices(const Mat_4& a, const Mat_4& b) {
+  return a + b;    
+}
+
+template<typename VCM, 
+         typename GeomTraits, 
+         typename QuadricImpl>
+class GarlandHeckbert_cost_base
+{
+  public:
+
+  // Tells the main function of 'Edge_collapse' that these
+  // policies must call "initialize" and "update" functions.
+  typedef CGAL::Tag_true Update_tag;  
+
+  typedef VCM Vertex_cost_map;
+  typedef typename GeomTraits::FT FT;
+  
+  typedef Eigen::Matrix<FT, 4, 4, Eigen::DontAlign> Mat_4;
+  typedef Eigen::Matrix<FT, 4, 1> Col_4;
+  
+  typedef typename GeomTraits::Point_3 Point_3;
+  typedef typename GeomTraits::Vector_3 Vector_3;
+
+  GarlandHeckbert_cost_base() : m_cost_matrices() { } 
+  
+  GarlandHeckbert_cost_base(FT dm) 
+    : m_cost_matrices(), discontinuity_multiplier(dm) { } 
+  
+  GarlandHeckbert_cost_base(Vertex_cost_map vcm, FT dm) 
+    : m_cost_matrices(vcm), discontinuity_multiplier(dm) { } 
+
+  // the deriving implementation needs this to fulfill the concept 
+  GarlandHeckbert_cost_base get_cost() const {
+    return *this;
+  }
+
+  // initialize all quadrics 
+  template<typename TM,
+           typename VPM> 
+  void initialize(const TM& tmesh, const VPM& vpm, const GeomTraits& gt) const
+  {
+    // the graph library types used in the rest
+    typedef boost::graph_traits<TM> GraphTraits;
+    typedef typename GraphTraits::vertex_descriptor vertex_descriptor;
+    typedef typename GraphTraits::halfedge_descriptor halfedge_descriptor;
+    typedef typename GraphTraits::face_descriptor face_descriptor;
+
+    Mat_4 zero_mat = Mat_4::Zero();
+    
+    for(vertex_descriptor v : vertices(tmesh))
+    {
+      put(m_cost_matrices, v, zero_mat);
+    }
+
+    for(face_descriptor f : faces(tmesh))
+    {
+      if (f == GraphTraits::null_face())
+      {
+        continue;
+      }
+
+      const halfedge_descriptor h = halfedge(f, tmesh);
+
+      // get the three vertices of the triangular face
+      const Point_3 p = get(vpm, source(h, tmesh));
```

Comment:
These were Point_reference in the original code and they should remain so: you are copying points there.